### PR TITLE
doc(features): Highlight the non-blocking feature gating technique

### PR DESCRIPTION
**Pull request by epage**
We already discussed non-blocking gates but the language makes it sound like it was limited to `config.toml`.
Since I haven't been touching that, I had always overlooked that section.

This change brings the blocking / non-blocking decision front and center.
To support this, the later sections focus more on mechanisms (the gate) rather than on what is being done (new syntax for `cargo-features`).

I also feel this makes the content more scannable.

This is adapted from what I did for `[lints]` (see #12148).

### CODE REVIEW

File: src/cargo/core/features.rs [hunk 1]
```diff
@@ -9,29 +9,43 @@
...
-//! * New functionality that may not have an interface, or the interface has
-//!   not yet been designed, or for more complex features that affect multiple
-//!   parts of Cargo should use a new `-Z` flag.
+//! * Error when the feature is used without the gate
+//!   * Required if ignoring the feature violates the users intent in non-superficial ways
+//!   * A low-effort / safe way to protect the user from being broken if the format of the feature changes in
+//!     incompatible was (can be worked around)
+//!   * Good for: CLI (gate: `-Zunstable-options` or `-Z` if combined with other changes), `Cargo.toml` (gate: `cargo-features`)
+//! * Warn that the feature is ignored due to lack of the gate
+//!   * For if you could opt-in to the unimplemented feature on cargo today and cargo would
```
Comment (weihanglo) [inline src/cargo/core/features.rs:18]: nit: capitalized when referring to the project?

(and ditto throughout the PR)

```suggestion
//!   * For if you could opt-in to the unimplemented feature on Cargo today and Cargo would
```

**Review by weihanglo — APPROVED** [review]
Looks good. Feel free to address nits or `r=me`.

BTW, by “non-blocking feature gate” does it refer to a feature gate that won't bail out but warn/ignore?

**Comment by epage** [comment]
> Looks good. Feel free to address nits or `r=me`.
> 
> BTW, by “non-blocking feature gate” does it refer to a feature gate that won't bail out but warn/ignore?

Yes.  Didn't end up using that term in the PR.

**Comment by epage** [comment]
nits addressed

@bors r=weihanglo

**Pull request merged**